structs-in-rust/article.md

```diff
@@ -0,0 +1,366 @@
+# Structs in Rust
+Almost every programming language allows the creation of some kind of data structure (class, record, data) which can pack a group of things together. Rust is no exclusion with its struct.
+
+By the end of this article you will learn:
+- what a struct is
+- how to define and initialize a struct
+- what is a tuple struct
+- traits and traits deriving
+- struct methods
+
+Code from this post is available at [Gist](https://gist.github.com/sancho20021/d6faa25ef319db2442cdfbb0aef6b9a0).
+
+## What is a struct in Rust?
+Struct is a structure that holds multiple related values. Let's jump straight to our first example and define a Point — wrapper for two coordinates.
+```
+// [1]   [2]
+struct Point {
+// [3] [4]
+    x: i32,
+    y: i32,
+}
+
+// We
+// [1]: tell Rust that we want to define a new struct
+// [2]: named "Point"
+// [3]: that contains two fields with names "x" and "y"
+// [4]: both of types "i32"
+```
+Struct in Rust reflects the same idea as a struct in C or a class in Java. Consider the analogue of the above code in mentioned languages:
+```
+// C
+struct Point {
+    int x;
+    int y;
+};
+
+// Java
+class Point {
+    int x;
+    int y;
+}
+```
+
+### Unit struct
+Unit struct is a trivial example that contains zero fields:
+```
+struct A;
+```
+Don't be afraid if you see such strange struct definitions. You may wonder what is the purpose of unit structs when they don't contain any information. In reality, they are being used (see [std::fmt::Error](https://doc.rust-lang.org/std/fmt/struct.Error.html), [chrono::offset::Utc](https://docs.rs/chrono/0.4.19/chrono/offset/struct.Utc.html)), as they can have methods and different trait implementations. Both things I just mentioned will be discussed further in the article.
```

Comment:
This isn't very convincing, could you come up with a simple example?

Reply:
Ok. I think the example would require generics and of course traits. What do you think of placing the example after the `Traits` section?

Reply:
That feels too complicated for this. 🤔 Can it be textual -- a description of what would you do.

Reply:
I want to show smth like this:
```
mod library {
    pub trait ErrorLogger {
        fn error(error_code: i32);
    }

    pub fn function_with_errors<L: ErrorLogger>() {
        L::error(25);
    }
}

mod my_app {
    use super::library::*;

    // Logger 1
    struct PoliteLogger;

    impl ErrorLogger for PoliteLogger {
        fn error(error_code: i32) {
            println!(
                "Sorry for interrupting you, but an error with code {} occurred.",
                error_code
            );
        }
    }

    // Logger 2
    struct SimpleLogger;

    impl ErrorLogger for SimpleLogger {
        fn error(error_code: i32) {
            println!("Error: {}", error_code);
        }
    }

    fn option1() {
        function_with_errors::<PoliteLogger>();
    }

    fn option2() {
        function_with_errors::<SimpleLogger>();
    }
}
```

So do I need to describe the idea in words? But we assume the reader doesn't know that functions can take generic arguments.

Reply:
What about dropping unit structs? I can't come up with a simple explanation without leaving the reader with questions

Reply:
Yeah, that is possible as well. 👍

Reply:
I decided to remove it)